src/api/documents/document-ref.ts

```diff
@@ -0,0 +1,121 @@
+// Copyright 2021, Nitric Technologies Pty Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import { Struct } from 'google-protobuf/google/protobuf/struct_pb';
+import {
+  DocumentServiceClient,
+  DocumentGetRequest,
+  DocumentGetResponse,
+  Key,
+  Collection as DocCollection,
+  DocumentSetRequest,
+  DocumentDeleteRequest,
+} from '../../interfaces/document';
+import { CollectionRef } from './collection-ref';
+import { getKey } from './key';
+
+/**
+ * Document Ref
+ *
+ * Provides a Document Reference class.
+ * Used to create references to collections.
+ */
+export class DocumentRef<T extends { [key: string]: any }> {
+  private documentClient: DocumentServiceClient;
+  private key: Key;
+
+  constructor(
+    documentClient: DocumentServiceClient,
+    collection: DocCollection,
+    documentId: string
+  ) {
+    this.documentClient = documentClient;
+    this.key = getKey(collection, documentId);
+  }
+
+  /**
+   * Return the collection document reference value.
+   * @returns the collection document reference value, or null if not found
+   */
+  public async get() {
+    const request = new DocumentGetRequest();
+    request.setKey(this.key);
+
+    return new Promise<T>((resolve, reject) => {
+      this.documentClient.get(
+        request,
+        (error, response: DocumentGetResponse) => {
+          if (error) {
+            reject(error);
+          } else if (response.hasDocument()) {
+            const document = response.getDocument();
+            const content = document.getContent().toJavaScript() as T;
+
+            resolve(content);
+          } else {
+            resolve(null);
+          }
+        }
+      );
+    });
+  }
+
+  /**
+   * Set the document content for this document reference in the database. If the
+   * document does not exist an new item will be created, otherwise an
+   * existing document will be update with the new value.
+   * @param value content the document content to store (required)
+   */
+  public async set(value: Partial<T>) {
```

Comment:
I think Partial works well where we want to allow merging, but could potentially cause errors where users want to create or overwrite.

For Partial, I think users are likely to mark optional fields for types, in the cases of set of overwrite, of course we need partial for merge. I know that firestore adds merge as a flag on their set method, but I think separating out the concept into a separate method signature would provide better application level type safety for out typescript users. (although it would be redundant for our javascript users.

Thoughts?

Reply:
If we separate merging into another update method then I totally agree. I will remove the partial